File: lib/models/service.dart
enum ServiceCategory {
  lash,
  pmu,
  nail,
  hair,
  skin,
  other,
}

extension ServiceCategoryExtension on ServiceCategory {
  String get displayName {
    switch (this) {
      case ServiceCategory.lash:
        return '睫毛';
      case ServiceCategory.pmu:
        return '半永久化妝';
      case ServiceCategory.nail:
        return '美甲';
      case ServiceCategory.hair:
        return '美髮';
      case ServiceCategory.skin:
        return '美容';
      case ServiceCategory.other:
        return '其他';
    }
  }

  String toJson() {
    return toString().split('.').last;
  }

  static ServiceCategory fromString(String value) {
    return ServiceCategory.values.firstWhere(
      (category) => category.toString().split('.').last == value,
      orElse: () => ServiceCategory.other,
    );
  }
}

class Service {
  final String id;
  final String businessId;
  final String name;
  final ServiceCategory category;
  final int duration; // 分鐘
  final int revisitPeriod; // 天
  final double price;
  final double profit;
  final String? description;
  final bool isArchived;
  final DateTime createdAt;
  final DateTime updatedAt;

  Service({
    required this.id,
    required this.businessId,
    required this.name,
    required this.category,
    required this.duration,
    required this.revisitPeriod,
    required this.price,
    required this.profit,
    this.description,
    this.isArchived = false,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Service.fromJson(Map<String, dynamic> json) {
    return Service(
      id: json['id'] as String,
      businessId: json['business_id'] as String,
      name: json['name'] as String,
      category: ServiceCategoryExtension.fromString(json['category'] as String),
      duration: json['duration'] as int,
      revisitPeriod: json['revisit_period'] as int,
      price: (json['price'] as num).toDouble(),
      profit: (json['profit'] as num).toDouble(),
      description: json['description'] as String?,
      isArchived: json['is_archived'] as bool? ?? false,
      createdAt: DateTime.parse(json['created_at'] as String),
      updatedAt: DateTime.parse(json['updated_at'] as String),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'business_id': businessId,
      'name': name,
      'category': category.toJson(),
      'duration': duration,
      'revisit_period': revisitPeriod,
      'price': price,
      'profit': profit,
      'description': description,
      'is_archived': isArchived,
      'created_at': createdAt.toIso8601String(),
      'updated_at': updatedAt.toIso8601String(),
    };
  }

  Service copyWith({
    String? id,
    String? businessId,
    String? name,
    ServiceCategory? category,
    int? duration,
    int? revisitPeriod,
    double? price,
    double? profit,
    String? description,
    bool? isArchived,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return Service(
      id: id ?? this.id,
      businessId: businessId ?? this.businessId,
      name: name ?? this.name,
      category: category ?? this.category,
      duration: duration ?? this.duration,
      revisitPeriod: revisitPeriod ?? this.revisitPeriod,
      price: price ?? this.price,
      profit: profit ?? this.profit,
      description: description ?? this.description,
      isArchived: isArchived ?? this.isArchived,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }
} 

File: backend/internal/models/service.go
package models

import (
	"time"

	"gorm.io/gorm"
)

// ServiceCategory constants
const (
	ServiceCategoryLash  = "lash"
	ServiceCategoryPMU   = "pmu"
	ServiceCategoryNail  = "nail"
	ServiceCategoryHair  = "hair"
	ServiceCategorySkin  = "skin"
	ServiceCategoryOther = "other"
)

// Service represents the service model
type Service struct {
	ID            string         `gorm:"primaryKey;type:varchar(36)" json:"id"`
	BusinessID    string         `gorm:"type:varchar(36);not null;index" json:"business_id"`
	Name          string         `gorm:"type:varchar(100);not null" json:"name"`
	Category      string         `gorm:"type:varchar(20);not null" json:"category"`
	Duration      int            `gorm:"not null" json:"duration"`       // 分鐘
	RevisitPeriod int            `gorm:"not null" json:"revisit_period"` // 天
	Price         float64        `gorm:"type:decimal(10,2);not null" json:"price"`
	Profit        float64        `gorm:"type:decimal(10,2);not null" json:"profit"`
	Description   *string        `gorm:"type:text" json:"description,omitempty"`
	IsArchived    bool           `gorm:"default:false" json:"is_archived"`
	CreatedAt     time.Time      `json:"created_at"`
	UpdatedAt     time.Time      `json:"updated_at"`
	DeletedAt     gorm.DeletedAt `gorm:"index" json:"-"`

	// Relationships
	Business *Business `gorm:"foreignKey:BusinessID" json:"business,omitempty"`
}

// TableName specifies the table name for Service model
func (Service) TableName() string {
	return "services"
}

// BeforeCreate hook to generate UUID for service ID
func (s *Service) BeforeCreate(tx *gorm.DB) error {
	if s.ID == "" {
		s.ID = generateUUID()
	}
	return nil
}

// CreateServiceRequest represents the create service request payload
type CreateServiceRequest struct {
	BusinessID    string  `json:"business_id" binding:"required"`
	Name          string  `json:"name" binding:"required,min=2,max=100"`
	Category      string  `json:"category" binding:"required"`
	Duration      int     `json:"duration" binding:"required,min=1"`
	RevisitPeriod int     `json:"revisit_period" binding:"required,min=0"`
	Price         float64 `json:"price" binding:"required,min=0"`
	Profit        float64 `json:"profit" binding:"required,min=0"`
	Description   string  `json:"description,omitempty"`
	IsArchived    bool    `json:"is_archived,omitempty"`
}

// UpdateServiceRequest represents the update service request payload
type UpdateServiceRequest struct {
	Name          string  `json:"name" binding:"omitempty,min=2,max=100"`
	Category      string  `json:"category,omitempty"`
	Duration      int     `json:"duration" binding:"omitempty,min=1"`
	RevisitPeriod int     `json:"revisit_period" binding:"omitempty,min=0"`
	Price         float64 `json:"price" binding:"omitempty,min=0"`
	Profit        float64 `json:"profit" binding:"omitempty,min=0"`
	Description   string  `json:"description,omitempty"`
	IsArchived    *bool   `json:"is_archived,omitempty"`
}

// ToModel converts CreateServiceRequest to Service model
func (r *CreateServiceRequest) ToModel() *Service {
	service := &Service{
		BusinessID:    r.BusinessID,
		Name:          r.Name,
		Category:      r.Category,
		Duration:      r.Duration,
		RevisitPeriod: r.RevisitPeriod,
		Price:         r.Price,
		Profit:        r.Profit,
		IsArchived:    r.IsArchived,
	}

	if r.Description != "" {
		service.Description = &r.Description
	}

	return service
}

// IsActive returns true if the service is not archived
func (s *Service) IsActive() bool {
	return !s.IsArchived
}

// ServiceCategoryResponse represents service category information
type ServiceCategoryResponse struct {
	Key         string `json:"key"`
	DisplayName string `json:"display_name"`
}

// GetServiceCategories returns all available service categories
func GetServiceCategories() []ServiceCategoryResponse {
	return []ServiceCategoryResponse{
		{Key: ServiceCategoryLash, DisplayName: "睫毛"},
		{Key: ServiceCategoryPMU, DisplayName: "半永久化妝"},
		{Key: ServiceCategoryNail, DisplayName: "美甲"},
		{Key: ServiceCategoryHair, DisplayName: "美髮"},
		{Key: ServiceCategorySkin, DisplayName: "美容"},
		{Key: ServiceCategoryOther, DisplayName: "其他"},
	}
}
